### DISCUSSION TITLE
MAUI should have a RichTextBox equivalent rich text editor

### DISCUSSION THREAD

**Discussion by IcicleTrepan** [dotnet/maui#8087] [Ideas]
I'd really like to use .NET MAUI to build a cross platform desktop application that works in Windows as well as Mac. 

One of the biggest issues I've come across so far while attempting to port an app from WPF to MAUI is that there's no equivalent to RichTextBox. This type of control is integral to the application I'm working on and unfortunately, this type of control can't even be purchased from a 3rd party at this moment in time as it doesn't exist.

Please consider adding an equivalent to RichTextBox going forward. Not everyone is using this for mobile alone, please don't forget us desktop people!

**Comment by dustinhorne** [comment]
I'll second this.  I'd love to work natively in MAUI but I need a rich text editor for my project.  As a workaround, I'm just building a hybrid app and using Blazor and Quill inside my MAUI app.  It works but isn't necessarily ideal.

**Comment by giuseppenovielli** [comment]
@IcicleTrepan As a workaround, you can use WebView control and loading opensource web RichTextEditor, as example 
https://ckeditor.com/

**Reply by PerfectlyFineCode** [reply]
I'm sorry, but what?
Are we supposed to use webview for a lousy richtextbox?
Microsoft seems to be over their head if that's the plan.

> @IcicleTrepan As a workaround, you can use WebView control and loading opensource web RichTextEditor, as example https://ckeditor.com/

**Reply by giuseppenovielli** [reply]
No, I wrote to you if it can be useful. 
It is clear that the best approach would be a native RichTextEditor!

**Comment by xtuzy** [comment]
I get these infor from Google and Github.
1. If you only need simple Bold/Italic like this:
![image](https://user-images.githubusercontent.com/17793881/194824755-669cc781-56a9-4388-888f-42b15ae4a6dd.png)
you can see https://github.com/TimPurdum/RichTextEditor

2. If you need many function of richtextedit, you can use webview or binding native library, 
   - native library like https://github.com/wasabeef/richeditor-android have many star, but it have died. you can find little star still be develop, like https://github.com/1gravity/Android-RTEditor, if you want use native library, you need  find one fit your requirement. Another, binding ios project is troublesome
   - webview, you can see https://github.com/Blazored/TextEditor, it use quilljs

ReactNative:
- [react-native-cn-richtext-editor](https://github.com/imnapo/react-native-cn-richtext-editor) have died, it advice [react-native-cn-quill](https://github.com/imnapo/react-native-cn-quill);
- [react-native-draftjs](https://github.com/DaniAkash/react-native-draftjs) have died, it use facebook [draft-js](https://github.com/facebook/draft-js), current draft-js have died, facebook develop [lexical](https://github.com/facebook/lexical), you can see reactnative developer want it [discussions](https://github.com/facebook/lexical/discussions/2410)
- [react-native-rich-editor](https://github.com/wxik/react-native-rich-editor), also use webview, like died

Flutter:
 -  [flutter-quill](https://github.com/singerdmx/flutter-quill) very cool, it let me feel flutter is light web
 - [html-editor-enhanced](https://github.com/tneotia/html-editor-enhanced)

**Comment by pvladov** [comment]
Another solution would be to use a third party cross-platform rich text editor control (not MAUI specific). For example, the [NOV Rich Text Editor control](https://www.nevron.com/products-open-vision-nov-rich-text-editor-control-overview.aspx) works on the following platforms from single codebase: Windows (WinForms and WPF), macOS (Xamarin.Mac) and web (Blazor WebAssembly). You write the code once in a .NET Standard assembly using the NOV Rich Text Editor and then the same code compiles and runs on all supported platforms. Then on each platform you just need a simple executable project that initializes NOV and shows the main widget of the application.

More information is available on its website:
[NOV Rich Text Editor Control for .NET WinForms, WPF, Xamarin.Mac and Blazor](https://www.nevron.com/products-open-vision-nov-rich-text-editor-control-overview.aspx)

Online demo using Blazor WebAssembly is available on:
[NOV WebAssembly Examples](https://blazorexamples.nevron.com/)

Disclosure:
I'm one of the developers working on the NOV cross-platform controls.

**Reply by dustinhorne** [reply]
I tried to get to the demo but it's apparently not mobile friendly. 
![SmartSelect_20221012_065737_Edge](https://user-images.githubusercontent.com/4991606/195336786-d5f97ae3-1ae0-45e7-8e5f-0440181807dc.jpg)

**Reply by pvladov** [reply]
Hi  @dustinhorne ,

The current version of the NOV cross-platform controls is designed for the desktop - Windows (WinForms and WPF), macOS (Xamarin.Mac) and web (Blazor WebAssembly).

**Reply by dustinhorne** [reply]
Ahh I'll stick with Blazor/Quill in Maui.  Using Bulma the whole app is responsive, I can reuse the UI for an identical web experience,  and Quill's excellent Delta system makes generating other doc formats as easy as figuring out the specs.

**Comment by joveaaron** [comment]
Bumping for awareness!

**Reply by parentelement** [reply]
I've actually started building one out.  I've separated it into an API and a frontend.  I'm targeting Avalonia to start with but the goal is to have it functional for Maui and Uno as well which is why I abstracted the API and Platform bits. It's a lot of work and slow going so if anyone wanted to help build it out, especially the Maui pieces I'd welcome the assist.  Here's an early video of the progress.  It's a fair bit further along now including some basic HTML and OpenXML export.  There are still a few bugs I need to address and some features to add.

https://youtu.be/KBmK3IvM9G8?si=uyjwyTjInIPhIV0r

**Comment by baaaaif** [comment]
not free... but https://www.telerik.com/maui-ui/richtexteditor

**Comment by meyerdominik** [comment]
See #29016
